### PR TITLE
#167735041 Set up Profile Page endpoints

### PR DESCRIPTION
**Pull request by elikeyz**
**What does this PR do?**

  Set up the endpoints for the user profile page

  **Description of Task to be completed?**

  - create an endpoint for the user to be able to see his/her profile
  information
  - create an endpoint for the user to be able to update his/her profile
  information
  - write tests for these endpoints
  - update the swagger documentation
  - update the readme
  - update the Dockerfile

  **How should this be manually tested?**

  - Clone this repo
  - Run `npm install` to install dependencies
  - Run `npm install -g sequelize-cli` to install sequelize CLI on your computer
  - Set up your database environmental variables and run `sequelize db:migrate` to run migrations
  - Run `npm start` to start the app
  - Then make a get request and a patch request respectively to `localhost:3000/api/v1/users/profile`

  **What are the relevant stories?**

  [#167735041](https://www.pivotaltracker.com/n/projects/2382051)

  **Screenshots (if appropriate)**

Screenshot (61)

Screenshot (62)

### CODE REVIEW

File: src/controllers/usersController.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+ * @param {object} user - user auth token payload
+ * @param {object} res - response object
+ * @return {object} user - return object containing status and data
+ */
+  static async getProfile({ user }, res) {
+    try {
+      const { id } = user;
+      const userData = await UsersService.getProfile(id);
+
+      res.status(200).json({ status: 'success', data: userData });
```
Comment (prieste) [inline src/controllers/usersController.js]: there's a utility function for sending responses in the utils folder, please utilizes it

File: src/controllers/usersController.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+ * @param {object} user - user auth token payload
+ * @param {object} res - response object
+ * @return {object} user - return object containing status and data
+ */
+  static async updateProfile({ body, user }, res) {
+    try {
+      const { id } = user;
+      const updatedData = await UsersService.updateProfile(id, body);
+
+      res.status(200).json({ status: 'success', data: updatedData });
```
Comment (prieste) [inline src/controllers/usersController.js]: here's a utility function for sending responses in the utils folder, please utilizes it

**Review by prieste — COMMENTED** [review]
Nice job

File: src/routes/users.js [hunk 3]
```diff
@@ -0,0 +1,12 @@
+import express from 'express';
+import UsersController from '../controllers/usersController';
+import authMiddleware from '../middlewares/auth';
+import { profileSchema } from '../validation/schemas';
+import validate from '../validation/validator';
+
+const router = express.Router();
+
+router.get('/me', authMiddleware, UsersController.getProfile);
+router.patch('/me', authMiddleware, validate(profileSchema, 'body'), UsersController.updateProfile);
```
Comment (cleave3) [inline src/routes/users.js]: Instead of using '/me' ,  i think  '/profile' is a more descriptive name for this  route

File: src/controllers/usersController.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+ * @param {object} user - user auth token payload
+ * @param {object} res - response object
+ * @return {object} user - return object containing status and data
+ */
+  static async getProfile({ user }, res) {
+    try {
+      const { id } = user;
+      const userData = await UsersService.getProfile(id);
+
+      Response.success(res, userData);
```
Comment (cleave3) [inline src/controllers/usersController.js]: Destruction success and badRequest from Response. That is, `const { success, badRequest} = Response`
So you can have `sucess(res, userData)`.  You also apply this in the other places where i applies. I This think this will make your code cleaner

File: README.md [hunk 5]
```diff
@@ -35,7 +35,11 @@ docker run -p 3000:3000 iceman-backend
 <table>
 <tr><th>HTTP VERB</th><th>ENDPOINTS</th><th>DESCRIPTION</th></tr>
 <tr><td>POST</td><td>/api/v1/auth/signup</td><td>Creates user account</td></tr>
-<tr><td>POST</td><td>/api/v1/auth/signup</td><td>Login user account</td></tr>
+<tr><td>POST</td><td>/api/v1/auth/login</td><td>Logs user in</td></tr>
+<tr><td>GET</td><td>/api/v1/users/me</td><td>Gets user profile information</td></tr>
```
Comment (Annmary12) [inline README.md]: Why 'me' in your route?

Comment (elikeyz) [reply README.md]: Sorry, I forgot to change that to `/profile`. I'll do that asap.

File: src/controllers/usersController.js [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    try {
+      const { id } = user;
+      const updatedData = await UsersService.updateProfile(id, body);
+
+      success(res, updatedData);
+    } catch ({ message: error }) {
+      badRequest(res, error);
+    }
+  }
+}
```
Comment (Annmary12) [inline src/controllers/usersController.js]: We have `AuthController`, please ultilize that instead of creating your own controller.

Comment (elikeyz) [reply src/controllers/usersController.js]: Ok, noted

File: src/docs/docs.yaml [hunk 7]
```diff
@@ -197,6 +287,96 @@ definitions:
...
+            type: integer
+            nullable: true
+          gender:
+            type: string
+            nullable: true
+          date_of_birth:
+            type: string
+            format: date
+            nullable: true
+          department:
```
Comment (Annmary12) [inline src/docs/docs.yaml]: Please remove `manager_id` and `department`

File: src/docs/docs.yaml [hunk 8]
```diff
@@ -197,6 +287,96 @@ definitions:
...
+            type: string
+            nullable: true
+          preferred_language:
+            type: string
+            nullable: true
+          residential_address:
+            type: string
+            nullable: true
+          preferred_currency:
```
Comment (Annmary12) [inline src/docs/docs.yaml]: do we have `residential_address` and `preferred_currency` in our user model?

File: src/docs/docs.yaml (hunk 7 above)
Comment (elikeyz) [reply src/docs/docs.yaml]: Alright

File: src/docs/docs.yaml (hunk 8 above)
Comment (elikeyz) [reply src/docs/docs.yaml]: Yes, we do

File: src/middlewares/auth.js [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+const jwtSecret = process.env.JWTSECRET;
+export default async (req, res, next) => {
+  const token = req.header('token') || req.header('Authorization');
+  if (!token) return res.status(401).json({ status: 'error', error: 'Access Denied, No token provided' });
+  try {
+    const payload = await jwt.verify(token, jwtSecret);
+    req.user = payload;
+    next();
+  } catch (error) {
+    res.status(400).json({ status: 'error', error: 'Access Denied, Invalid token' });
```
Comment (Annmary12) [inline src/middlewares/auth.js]: Replace line 19 with our helper function

File: src/docs/docs.yaml (hunk 8 above)
Comment (elikeyz) [reply src/docs/docs.yaml]: I added it to the model and migrations in my branch

File: src/middlewares/auth.js (hunk 9 above)
Comment (elikeyz) [reply src/middlewares/auth.js]: Ok

File: src/docs/docs.yaml (hunk 7 above)
Comment (elikeyz) [reply src/docs/docs.yaml]: Do I also remove it from the model and migrations?

File: src/controllers/usersController.js (hunk 6 above)
Comment (elikeyz) [reply src/controllers/usersController.js]: Do I also utilize `AuthService` as well?

File: src/docs/docs.yaml (hunk 7 above)
Comment (Annmary12) [reply src/docs/docs.yaml]: Yes, we are replacing that with `role_id`

File: src/docs/docs.yaml (hunk 8 above)
Comment (Annmary12) [reply src/docs/docs.yaml]: Okay

File: src/controllers/usersController.js (hunk 6 above)
Comment (Annmary12) [reply src/controllers/usersController.js]: Yes Please..

File: src/docs/docs.yaml (hunk 7 above)
Comment (elikeyz) [reply src/docs/docs.yaml]: Ok

File: src/routes/users.js [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+import authMiddleware from '../middlewares/auth';
+import { profileSchema } from '../validation/schemas';
+import validate from '../validation/validator';
+
+const router = express.Router();
+
+router.get('/profile', authMiddleware, UsersController.getProfile);
+router.patch('/profile', authMiddleware, validate(profileSchema, 'body'), UsersController.updateProfile);
+
+export default router;
```
Comment (Annmary12) [inline src/routes/users.js]: Please move this also to the Auth route file, don't create yours.

File: src/controllers/usersController.js (hunk 6 above)
Comment (elikeyz) [reply src/controllers/usersController.js]: Alright

File: src/services/usersService.js [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+      'preferred_language',
+      'residential_address',
+      'preferred_currency',
+      'date_of_birth',
+    ]);
+    const result = await User.update(importantFields, { returning: true, where: { id: userId } });
+    const [, [{ dataValues: updatedData }]] = result;
+    return Helper.omitFields(updatedData, ['password']);
+  }
+}
```
Comment (Annmary12) [inline src/services/usersService.js]: Also utilize the Auth service here.

File: src/routes/users.js (hunk 10 above)
Comment (elikeyz) [reply src/routes/users.js]: Alright, noted

File: src/services/usersService.js (hunk 11 above)
Comment (elikeyz) [reply src/services/usersService.js]: Ok

File: src/controllers/authController.js [hunk 12]
```diff
@@ -89,9 +91,41 @@ export default class AuthController {
...
+ * @return {object} user - return object containing status and data
+ */
+  static async getProfile({ user }, res) {
+    try {
+      const { id } = user;
+      const userData = await AuthService.getProfile(id);
+
+      success(res, userData);
+    } catch ({ message: error }) {
+      badRequest(res, error);
```
Comment (Annmary12) [inline src/controllers/authController.js]: Please why do we have `res` in the catch block?

File: package.json [hunk 13]
```diff
@@ -9,8 +9,8 @@
     "server": "nodemon --watch src --exec babel-node src/server.js",
     "start": "babel-node src/server.js",
     "lint": "eslint ./src --fix",
-    "test": "cross-env NODE_ENV=test nyc --reporter=html --reporter=text mocha --timeout 500000 src/test/*.js --require @babel/register --exit",
-    "test-dev": "npm run db:migration && nyc --reporter=html --reporter=text cross-env NODE_ENV=test mocha --timeout 500000 src/test/*.js --require @babel/register && cross-env NODE_ENV=test  npm run undo:migration --exit",
+    "test": "cross-env NODE_ENV=test nyc --reporter=html --reporter=text mocha --timeout 500000 src/tests/*.js --require @babel/register --exit",
+    "test-dev": "npm run db:migration && nyc --reporter=html --reporter=text cross-env NODE_ENV=test mocha --timeout 500000 src/tests/*.js --require @babel/register && cross-env NODE_ENV=test  npm run undo:migration --exit",
```
Comment (Annmary12) [inline package.json]: Please remove the `test-dev` script.

File: src/routes/auth.js [hunk 14]
```diff
@@ -11,5 +14,7 @@ router.patch('/reset_password/:token', validate(passwordResetSchema, 'body'), Au
 router.post('/signup', validate(signUpSchema, 'body'), AuthController.signupUser);
 router.get('/verify', AuthController.verifyUser);
 router.post('/resend_verification_link', validate(verifyEmail, 'body'), AuthController.resendVerification);
+router.get('/profile', verifyUser, AuthController.getProfile);
+router.patch('/profile', verifyUser, validate(profileSchema, 'body'), AuthController.updateProfile);
```
Comment (Annmary12) [inline src/routes/auth.js]: Instead of having `AuthController.updateProfile`, we can destructure here

`const { updateProfile, getProfile } = AuthController`

```
router.get('/profile', verifyUser, getProfile);
router.patch('/profile', verifyUser, validate(profileSchema, 'body'), updateProfile);
```

Please fix the rest, signup and the rest.

File: package.json (hunk 13 above)
Comment (elikeyz) [reply package.json]: Alright

Comment (elikeyz) [reply package.json]: Done

File: src/migrations/20190815191222-create-user.js [hunk 15]
```diff
@@ -42,6 +42,30 @@ module.exports = {
       type: Sequelize.BOOLEAN,
       defaultValue: false
     },
+    role: {
```
Comment (Annmary12) [inline src/migrations/20190815191222-create-user.js]: line 45 suppose to be `roleId` instead of `role`. Reach out to @amoskeyz for further explanation.

File: src/models/User.js [hunk 16]
```diff
@@ -38,6 +38,30 @@ module.exports = (sequelize, DataTypes) => {
         type: DataTypes.BOOLEAN,
         defaultValue: false,
       },
+      role: {
```
Comment (Annmary12) [inline src/models/User.js]: ine 41 suppose to be `roleId` instead of `role`. Reach out to @amoskeyz for further explanation.

File: src/routes/auth.js [hunk 17]
```diff
@@ -1,20 +1,35 @@
...
-import { signUpSchema, passwordResetSchema, verifyEmail } from '../validation/schemas';
 import validate from '../validation/validator';
 
 const router = express.Router();
+const {
+  loginUser,
+  forgotPassword,
+  resetPassword,
+  signupUser,
+  verifyUser: verifyNewUser,
```
Comment (Annmary12) [inline src/routes/auth.js]: Why did you assign `verifyUser` to  `verifyNewUser`?

File: src/routes/auth.js [hunk 18]
```diff
@@ -1,20 +1,35 @@
...
 router.get('/google/callback', PassportController.callback('google'));
-router.post('/forgot_password', AuthController.forgotPassword);
-router.patch('/reset_password/:token', validate(passwordResetSchema, 'body'), AuthController.resetPassword);
-router.post('/signup', validate(signUpSchema, 'body'), AuthController.signupUser);
-router.get('/verify', AuthController.verifyUser);
-router.post('/resend_verification_link', validate(verifyEmail, 'body'), AuthController.resendVerification);
+router.post('/forgot_password', forgotPassword);
+router.patch('/reset_password/:token', validate(passwordResetSchema, 'body'), resetPassword);
+router.post('/signup', validate(signUpSchema, 'body'), signupUser);
+router.get('/verify', verifyNewUser);
```
Comment (Annmary12) [inline src/routes/auth.js]: Why can't it use `verifyUser` at line 30?

File: src/services/authService.js [hunk 19]
```diff
@@ -23,8 +23,7 @@ export default class AuthService {
   static async login(email, password) {
     const result = await User.findOne({ where: { email, is_verified: true } });
 
-
-    if (!result) error('The account does not exists or not yet verified');
+    if (!result) throw new Error('The account does not exist or not yet verified');
```
Comment (Annmary12) [inline src/services/authService.js]: Why did you remove `error`? ideally error function throws an error, please revert it back to the way it was.

File: src/services/authService.js [hunk 20]
```diff
@@ -141,7 +140,7 @@ export default class AuthService {
 
     const isUser = await User.findOne({ where: { id: isExpire.id } });
 
-    if (!isUser) error('User not find');
+    if (!isUser) throw new Error('User not found');
```
Comment (Annmary12) [inline src/services/authService.js]: Please take it back to the initial state.

Comment (Annmary12) [reply src/services/authService.js]: You can leave `found`

File: src/services/authService.js [hunk 21]
```diff
@@ -176,4 +175,34 @@ export default class AuthService {
...
+    return Helper.omitFields(user, ['password']);
+  }
+
+  /**
+* @param {number} userId - ID of logged in user
+* @param {object} profileDetails - object containing details to be changed
+* @return {object} - object containing updated user profile information
+*/
+  static async updateProfile(userId, profileDetails) {
+    const importantFields = Helper.pickFields(profileDetails, [
```
Comment (Annmary12) [inline src/services/authService.js]: Why are you picking importantFields?

File: src/services/authService.js [hunk 22]
```diff
@@ -176,4 +175,34 @@ export default class AuthService {
...
+      'last_name',
+      'gender',
+      'preferred_language',
+      'residential_address',
+      'preferred_currency',
+      'date_of_birth',
+    ]);
+    const result = await User.update(importantFields, { returning: true, where: { id: userId } });
+    const [, [{ dataValues: updatedData }]] = result;
+    return Helper.omitFields(updatedData, ['password']);
```
Comment (Annmary12) [inline src/services/authService.js:226]: Leave a line after line 205

File: src/tests/auth.test.js [hunk 23]
```diff
@@ -29,24 +29,25 @@ const user2 = {
...
+    notVerifiedUser = await TestHelper.createUser({
+      ...user,
+      email: 'user2@gmail.com',
     });
+  });
 
+  afterEach((done) => {
+    TestHelper.destroyModel('User');
+    done();
+  });
```
Comment (Annmary12) [inline src/tests/auth.test.js]: Why did you move this test from `before` to `afterEach`?

File: src/tests/users.test.js [hunk 24]
```diff
@@ -0,0 +1,143 @@
...
+      res.body.data.should.have.property('is_verified');
+      res.body.data.should.have.property('role');
+      res.body.data.should.have.property('gender').eql('Male');
+      res.body.data.should.have.property('date_of_birth');
+      res.body.data.should.have.property('preferred_language').eql('English');
+      res.body.data.should.have.property('preferred_currency').eql('Nigerian Naira (NGN)');
+      res.body.data.should.have.property('residential_address').eql('Benin City, Nigeria');
+    });
+  });
+});
```
Comment (Annmary12) [inline src/tests/users.test.js]: I believe we have `auth.test.js` file, please don't abstract yours to a separate file.

File: src/routes/auth.js (hunk 18 above)
Comment (elikeyz) [reply src/routes/auth.js]: `verifyUser` is also the name of the middleware that checks for auth token in the header

File: src/routes/auth.js (hunk 17 above)
Comment (elikeyz) [reply src/routes/auth.js]: `verifyUser` was already declared at the top as the middleware that checks for auth token in the header

File: src/models/User.js (hunk 16 above)
Comment (elikeyz) [reply src/models/User.js]: Alright

File: src/migrations/20190815191222-create-user.js (hunk 15 above)
Comment (elikeyz) [reply src/migrations/20190815191222-create-user.js]: Alright

File: src/routes/auth.js (hunk 14 above)
Comment (elikeyz) [reply src/routes/auth.js]: Ok, I'll do that

File: src/services/authService.js (hunk 20 above)
Comment (elikeyz) [reply src/services/authService.js]: Ok

File: src/services/authService.js (hunk 21 above)
Comment (elikeyz) [reply src/services/authService.js]: I didn't know about `.optional()` in `@hapi/joi` then, I'll change that now.

File: src/services/authService.js (hunk 22 above)
Comment (elikeyz) [reply src/services/authService.js:226]: Ok

File: src/tests/users.test.js (hunk 24 above)
Comment (elikeyz) [reply src/tests/users.test.js]: Alright

File: src/tests/auth.test.js (hunk 23 above)
Comment (elikeyz) [reply src/tests/auth.test.js]: This was to prevent the test data from affecting other tests that would run after it

Comment (elikeyz) [reply src/tests/auth.test.js]: I did it while I was fixing an issue with the tests. Amos didn't properly use hooks to seed his test data, so some of his tests were failing when tests that ran before it altered the User table in the test database.

File: src/routes/auth.js (hunk 17 above)
Comment (Annmary12) [reply src/routes/auth.js]: This will change if you have rebased from development branch, the `verifyUser` middleware has changed to `auth`

File: src/services/authService.js (hunk 21 above)
Comment (Annmary12) [reply src/services/authService.js]: Why do you want to use `.optional() in @hapi/joi`?

File: src/tests/auth.test.js (hunk 23 above)
Comment (Annmary12) [reply src/tests/auth.test.js]: You cannot change things to make your own test pass instead fix the test itself, these are kind of changes that messes up with our application.

File: src/services/authService.js (hunk 21 above)
Comment (elikeyz) [reply src/services/authService.js]: So that the user doesn't necessarily have to enter all the details in the schema if he/she is not updating everything. For example, if the user is just updating gender, he shouldn't have to type everything else as well.

File: src/routes/auth.js (hunk 17 above)
Comment (elikeyz) [reply src/routes/auth.js]: Ok

File: src/tests/auth.test.js (hunk 23 above)
Comment (elikeyz) [reply src/tests/auth.test.js]: Alright

File: src/services/authService.js (hunk 21 above)
Comment (Annmary12) [reply src/services/authService.js]: You don't need all this because you provide the fields you want the user to update and a user can practically update all his info except database `id,createdAt and updatedAt`

**Comment by Annmary12** [comment]
Resolve conflict on your PR

File: src/services/authService.js (hunk 21 above)
Comment (elikeyz) [reply src/services/authService.js]: Alright

**Comment by Annmary12** [comment]
@elikeyz resolve conflict on your PR
